New UI for task accordion

ardourApeX opened the pull request:
### Developer Name: Joy Gupta

### PR Number(s): 606

### Description:
Currently, the UI of the task accordion on the dashboard site does not seem to be satisfactory. It lacks in terms of UI, UX, color combinations, and readability. Additionally, pagination is implemented in the frontend. 

![image](https://github.com/Real-Dev-Squad/website-dashboard/assets/56365512/97be0344-1384-43da-8988-25026478d898)

### Requirements:
To enhance the UI and UX of the accordion, the following modifications/improvements should be implemented:

- The UI of each task card should match the one we have on the status site.
- Instead of using next and previous buttons, implement infinite scroll with pagination being handled on the backend site.
- Change the current API with the paginated one that we are using in `my-site`


### UI Fixes/Improvements:
- [x] Replace `assigned to` with `Created By`
- [x] Add `Priority` field to the right
- [x] Conditional rendering of `Progress` based on `Status`
- [x] Add `Type` field to the right
- [x] Implement infinite scroll and pagination in the backend, instead of handling it in the frontend
- [x] Use the tasks UI from my site as it also serves the same purpose
- [x] Change the current API with paginated one


### Issue Ticket Number:
- https://github.com/Real-Dev-Squad/website-dashboard/issues/601

### Backend Changes
- [ ] Yes
- [x] No

### Frontend Changes
- [x] Yes
- [ ] No

### Is Under Feature Flag
- [x] Yes
- [ ] No

### Database changes: 
- [ ]  Yes
- [x] No

### Breaking changes (If your feature is breaking/missing something please mention pending tickets): 
- [ ] Yes
- [x] No

### Deployment notes:
NA

### Tested in local
- [x] Yes
- [ ] No

### QA Instructions, Screenshots, Recordings
Test cases added :
![image](https://github.com/Real-Dev-Squad/website-dashboard/assets/56365512/51226f83-2c38-440a-aa24-68869552fc1a)
![image](https://github.com/Real-Dev-Squad/website-dashboard/assets/56365512/83dbb3ca-50fc-4d04-adfd-49684be235a3)

Test Stats:
Run `yarn test __tests__/user-details/task-duedate-hover.test.js ` and there you can see the test case running.
![image](https://github.com/Real-Dev-Squad/website-dashboard/assets/56365512/ed22c623-4c53-400d-b7c3-71fd8259c687)

Run `yarn test` and there you can see the test case running.
![image](https://github.com/Real-Dev-Squad/website-dashboard/assets/56365512/8719887b-caa8-4f15-bf39-c21c144fa18b)


### Working Proof
#### Functionality & Responsiveness

https://github.com/Real-Dev-Squad/website-dashboard/assets/56365512/c9c4bc8f-bd0c-497f-a24f-f0e8377c7726

ardourApeX commented:
Some of the cards may appear red even though the progress is 100%. This is because I've implemented the same logic that is currently used on the status side. Additionally, the progress bar will persist no matter what the status is. This is because this page does not allow anyone to change it; it is only in read-only mode. Also, hiding it wouldn't make sense because the Super User wouldn't be able to find out the task completion percentage.

Ajeyakrishna-k commented on users/details/utils.js:
```diff
@@ -36,3 +36,169 @@ function generateNoDataFoundSection(message) {
   container.appendChild(notFoundDiv);
 }
+
+function dateDiff(date1, date2, formatter) {
```
I think this function is used in other places also. Can you move this to the common utils and import it from there for both pages instead of duplicating the code?

ardourApeX replied:
Done

Ajeyakrishna-k commented:
> Some of the cards may appear red even though the progress is 100%. This is because I've implemented the same logic that is currently used on the status side. 

Status site task card colours are broken. Please don't implement same here.

Ajeyakrishna-k commented:
>Additionally, the progress bar will persist no matter what the status is. This is because this page does not allow anyone to change it; it is only in read-only mode. Also, hiding it wouldn't make sense because the Super User wouldn't be able to find out the task completion percentage.

We cannot update the status to DONE or VERIFIED unless the percentage is 100%. So even if you display that percentage it will be always 100% providing no extra value. That is the reason why it's proposed to remove the progress bar for other statuses in my-site and status site.

Ajeyakrishna-k commented:
Also please add more description. Currently it's not really clear what this PR does.

shubhamsigdar1 commented:
> * Implement infinite scroll and pagination in the backend, instead of handling it in the frontend

Hi @ardourApeX, before I review your pull request, I'd like to clarify a few things to avoid any confusion:

Where is your backend pull request?
Why is this point included in the pull request description?

Could you please provide more information about these points? This will help me better understand your pull request.

shubhamsigdar1 commented:
what you done in this pr...please give information(high level implementation in words please) in description
easy for us to review your pr as a reviewer.

ardourApeX commented:
> what you done in this pr...please give more information in description(high level implementation in words please else not understand what you done in this pr) easy for us to review your pr as a reviewer.
> 
> Where is your design doc???

While reviewing the issue ticket @iamitprakash suggested that design doc is not required as it just an UI update

ardourApeX commented:
> > * Implement infinite scroll and pagination in the backend, instead of handling it in the frontend
> 
> Hi @ardourApeX, before I review your pull request, I'd like to clarify a few things to avoid any confusion:
> 
> Where is your backend pull request? Why is this point included in the pull request description?
> 
> Could you please provide more information about these points? This will help me better understand your pull request.





> Also please add more description. Currently, it's not really clear what this PR does.

There are no backend changes related to this PR as I have just implemented an existing API. Also, I have updated both the PR and issue description. Let me know if anything else is required. Thank You!

ardourApeX commented:
> > Some of the cards may appear red even though the progress is 100%. This is because I've implemented the same logic that is currently used on the status side.
> 
> * Status site task card colours are broken. Please don't implement same here.
> * Issue for the right task colour in cards : [[Design Doc] -  Tasks card enhancement in status site website-status#931](https://github.com/Real-Dev-Squad/website-status/issues/931)

Given that we have not yet decided on the criteria for determining if a task will be considered overdue before the ETA, I have not implemented it at this time. The same we have discussed today.

ardourApeX commented:
> > Additionally, the progress bar will persist no matter what the status is. This is because this page does not allow anyone to change it; it is only in read-only mode. Also, hiding it wouldn't make sense because the Super User wouldn't be able to find out the task completion percentage.
> 
> We cannot update the status to DONE or VERIFIED unless the percentage is 100%. So even if you display that percentage it will be always 100% providing no extra value. That is the reason why it's proposed to remove the progress bar for other statuses in my-site and status site.

Done

sahsisunny reviewed (changes requested):
Test stats are useless showing coverage for mock data

ardourApeX commented:
> Test stats are useless showing coverage for mock data

Can you please elaborate it? What else is required from the tests?

sahsisunny commented on utils.js:
```diff
@@ -10,7 +10,7 @@ async function getSelfUser(endpoint = '/users/self') {
       const self_user = await res.json();
       if (res.status === 200) {
-        return self_user;
+        return self_user?.user || self_user;
```
Why `self_user?.user` ?

ardourApeX replied:
Because while running the test cases user details were present under the user key. On the other hand while running it locally, user details can be found inside self_user.

ardourApeX replied:
[Here is the](https://github.com/Real-Dev-Squad/website-dashboard/pull/606/files#diff-b6afa233efae4f2d689c44e64e59a981f7f2fe79f45c971f29a5040bfe677823L60) previous implementation, in which they were manually triggering the functions.
